src/crimson/os/seastore/seastore.cc

```diff
@@ -721,6 +722,12 @@ SeaStore::Shard::list_objects(CollectionRef ch,
 		  std::get<1>(ret) = std::get<1>(_ret);
 		  assert(limit >= next_objects.size());
 		  limit -= next_objects.size();
+		  LOG_PREFIX(SeaStore::list_objects);
+		  DEBUGT("got {} objects, left limit {}",
+		    t, next_objects.size(), limit);
+		  if (last && std::get<1>(ret) == pend) {
+		    std::get<1>(ret) = ghobject_t::get_max();
```

Comment:
See https://github.com/ceph/ceph/blob/main/src/os/ObjectStore.h#L678

I'm not sure which user logic will break if it is `pend`, not `ghobject_t::get_max()`, can you point it out?

It seems ObjectStore only requres it to be "next item sorts >= this value", `pend` should be good enough, right?

Reply:
> I'm not sure which user logic will break if it is `pend`, not `ghobject_t::get_max()`, can you point it out?

Backfill scanning rely on this to determine if the scan has reached the end.

Reply:
I see, it is (again) the issue that if returning `pend` and when the `limit` is also reached at the same time, user cannot determine whether the listing is complete. So `ghobject_t::get_max()` should be used as the indicator that the listing is complete.

I think this fix should be moved down to `FLTreeOnodeManager::list_onodes()` as it is a general issue. And add proper interface comments about when and why `ghobject_t::get_max()` will be returned.

Does it make sense?

Reply:
Yeah, will do.

Reply:
> I see, it is (again) the issue that if returning `pend` and when the `limit` is also reached at the same time, user cannot determine whether the listing is complete. So `ghobject_t::get_max()` should be used as the indicator that the listing is complete.
> 
> I think this fix should be moved down to `FLTreeOnodeManager::list_onodes()` as it is a general issue. And add proper interface comments about when and why `ghobject_t::get_max()` will be returned.
> 
> Does it make sense?

After a second thought, there seems no way for `OnodeManager` to know whether `ghobject_t::max()` needs to be returned, because it's only when the onode listing meets pend and it's the last range that we should return `ghobject_t::max()`.

Reply:
`FuturizedStore::list_objects()` and `OnodeManager::list_onodes()` are almost sharing the same interface signature. I think they should enforce the consistent behavior about how `std::tuple<std::vector<ghobject_t>, ghobject_t>` should be returned.

The interface design itself should be fine as long as it can also answer the two major questions with its the return values:
1. Whether this listing has reached the very last?
2. Whether this listing has reached the limit before the end parameter?

Reply:
Seems to me the current `OnodeManager::list_onodes()` can answer all the questions above:
1. When the 2nd ret is `ghobject_t::get_max()`.
2. When the length of the 1st ret is equal to the limit.
3. When the 2nd ret is equal to the input parameter `end`.

So probably it is a good idea to apply the same behavior to `SeaStore::list_objects()` if it is problematic?

Reply:
> So probably it is a good idea to apply the same behavior to `SeaStore::list_objects()` if it is problematic?

I don't quite follow, `SeaStore::list_objects()` is already doing this.

If we want `OnodeManager::list_onodes()` to have the same behaviour, it looks to me we need to move `SeaStore::get_objs_range()` and `struct col_obj_ranges_t` into `OnodeManager`, is this what you are suggesting?

Reply:
> I don't quite follow, `SeaStore::list_objects()` is already doing this.

Do you mean before or after this PR's changes?

Regardless of the specific implementations, if `SeaStore::list_objects()` is already able to answer all the 3 questions in the exact way of https://github.com/ceph/ceph/pull/53062#discussion_r1338333338, I'm not following why the interface design (if is correctly implemented) still cannot answer "whether the backfill scan has reached the end". Is there anything else missing from the interface definition?

Reply:
> > I don't quite follow, `SeaStore::list_objects()` is already doing this.
> 
> Do you mean before or after this PR's changes?
> 
> I want to make sure the interface design is correct first.
> 
> Regardless of the specific implementations, if `SeaStore::list_objects()` is already able to answer all the 3 questions in the exact way of [#53062 (comment)](https://github.com/ceph/ceph/pull/53062#discussion_r1338333338), I'm not following why the interface design (if is correctly implemented) still cannot answer "whether the backfill scan has reached the end". Is there anything else missing from the interface definition?

Sorry, I meant this PR has already made `SeaStore::list_objects()` answering those 3 questions, if we want `OnodeManager::list_onodes()` to answer those questions, we might need to move `SeaStore::get_objs_range()` and `struct col_obj_ranges_t` into `OnodeManager`, is this reasonable to you?

Reply:
> if we want `OnodeManager::list_onodes()` to answer those questions, we might need to move SeaStore::get_objs_range() and struct col_obj_ranges_t into OnodeManager, is this reasonable to you?

I'm not following the question here .. See https://github.com/ceph/ceph/pull/53062#discussion_r1338333338, they are abstracted from the behavior of `OnodeManager::list_onodes()` about how it answers the 3 questions listed.

Anyway, I'll verify whether `SeaStore::list_objects()` behaves in the intended way with this change.